Saas13.1 website slides quiz not joined sign in fle

fleodoo opened the pull request:
If applied, this commit will allow user to take a quiz (that is in preview) before being signed in
or having joined the cours. According to the current state of the user,
he will be redirected to signin/buy the course. After having signed in he is redirected to the quiz
where hes previous answers will be corrected.
For a buyable cours, when he has paid and the sale order is validated, the quiz will be corrected on
hes next visit to the quiz.

taskid:2045571

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -20,6 +20,46 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+    },
+    /**
+     * @private
+     * @param
```
Which param?

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -20,6 +20,46 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+     * @private
+     * @param
+     */
+    _createLoginRedirectUrlSale: function(){
```
This is more a getter than a "create".

So `_getLoginRedirectUrlSale`

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -49,6 +51,13 @@ var CourseJoinWidget = publicWidget.Widget.extend({
      */
     _onClickJoin: function (event) {
         var channelId = this.channelId || $(event.currentTarget).data('channel-id');
+        this.joinCours(channelId); 
```
`joinCourse`

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -57,7 +66,12 @@ var CourseJoinWidget = publicWidget.Widget.extend({
         }).then(function (data) {
             if (! data.error) {
-                location.reload();
+                if (self.quiz === true){
```
`if (self.quiz)`

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -73,33 +78,28 @@ odoo.define('website_slides.quiz', function (require) {
+                var courseJoinWidget = new CourseJoinWidget(self, self.channel.channelId,true);
+                return courseJoinWidget;
+            }).then(function(courseJoinWidget){
+                if(self.notMember && !self.publicUser){
```
Why is this in a `then` clause?

Looks like this code could be right after `var courseJoinWidget = new CourseJoinWidget(self, self.channel.channelId,true);`

fleodoo replied:
Because I wanted to make sur the widget was created before using .appendTo and .joinCourse. As I Understood JS is synchronous and you need to force asynchronous behavior with promises. Isn't this how it has to be done?

awa-odoo replied:
JS is asynchronous and you need to force synchronous behavior with promises*

Yes, it's indeed the case, but you don't need it here.

When you instantiate a widget, it calls the "init" method of that widget synchronously.
Then, when you use your `appendTo` just below, it calls the "start" method of the widget, that one is async.
So you can call appendTo right after instantiating the widget without any issue.

What you're doing correctly is putting all that in the  "then" phase of your parent widget, meaning it's DOM is ready.

Ideally, that whole block of code should be extracted in a sub-method, just like the `_renderX` above.
So you would have a `_renderJoinWidget()`.

If that's not clear, don't hesitate to ask :)

fleodoo replied:
I think it's clear thanks, I created the _renderJoinWidget() methode.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -73,33 +78,28 @@ odoo.define('website_slides.quiz', function (require) {
+                return courseJoinWidget;
+            }).then(function(courseJoinWidget){
+                if(self.notMember && !self.publicUser){
+                    if ($(".o_wslides_course_join_widget")[0]){
```
We try to avoid global selectors as much as possible.
You should only alter blocks inside your widget's element.

`self.$(".o_wslides_course_join_widget")` searches inside your widget and is what you should use here.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -219,6 +228,80 @@ odoo.define('website_slides.quiz', function (require) {
+        /**
+         * @private
+         */
+        _onClicksigninJoin: function (event) {
```
This looks very (very) similar to `_onClickJoinSale`

You should try to factorize the common parts.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -219,6 +228,80 @@ odoo.define('website_slides.quiz', function (require) {
+            else if (alert_code === 'public_user') {
+                message = _t('You must be logged to submit the quiz.');
+            }
+            this.$('.o_wslides_js_lesson_quiz_error span:first').html(message);
```
You should try to avoid `html()` as much as possible since it introduces XSS attacks points.
`.text()` should work just fine for this case.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -219,6 +228,80 @@ odoo.define('website_slides.quiz', function (require) {
+         * Get the answers filled in by the User
+         *
+         * @private
+         * @param 
```
No param

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -219,6 +228,80 @@ odoo.define('website_slides.quiz', function (require) {
+         * @param 
+         */
+        _getAnswers: function () {
+            var inputs = this.$('input[type=radio]:checked');
```
You should be able to use `.map()` to reduce code here.
Something like:

`return this.$('input[type=radio]:checked').map(function (element) {return $(element).val();});`

To check/test

fleodoo replied:
```

return jQuery.makeArray(this.$('input[type=radio]:checked').map(function (index, element) {
                return [$(element).val()];
            }));
```
It is working if I do this.

awa-odoo replied:
Can you try `return this.$('input[type=radio]:checked').map(function (element) {return $(element).val();}).get();` ?

The important difference being the `get`, sorry I forgot to put it in my first comment.

fleodoo replied:
this works:
```
return this.$('input[type=radio]:checked').map(function (index, element) {
    return parseInt($(element).val());
 }).get();
```

I need the parseint because otherwise it returns an array of strings and I work with ints later on and removing the 'index' argument of the function in the map causes an error.

awa-odoo replied:
Already a lot better, thanks for experimenting with it!

awa-odoo commented on addons/website_slides/static/src/xml/slide_quiz.xml:
```diff
@@ -46,27 +46,48 @@
-            <div t-if="widget.readonly &amp;&amp; !widget.publicUser &amp;&amp; widget.channel.channelEnroll == 'public'" class="o_wslides_course_join_widget">
+            <div t-if="widget.notMember &amp;&amp; !widget.publicUser &amp;&amp; widget.channel.channelEnroll == 'public'" class="alert alert-info d-flex align-items-center justify-content-between">
+                <div>
+                    <b class="h5 mb-0">join the course to take the quiz and verify your anwsers!</b>
```
"Join"

fleodoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -20,6 +20,46 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+                else{
+                    var url =_.str.sprintf('/shop/cart/update?product_id=%s&amp;express=1',(self.channel.productId));
+                    }
+                window.location=url
```
For some reason he was turned off, I turned it on again.

fleodoo commented:
@awa-odoo I think I finished my correction.

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -20,6 +20,28 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+     */
+    _onClickJoinSale: function (event) {
+        event.preventDefault();
+        var url = _.str.sprintf('/shop/cart/update?product_id=%s&amp;express=1', (this.channel.productId));
```
This part is common: `_.str.sprintf('/shop/cart/update?product_id=%s&amp;express=1', (this.channel.productId));`

You could put it in a variable like `baseShopUrl` and pass it to your `_getLoginRedirectUrlSale` method to avoid duplicating it.

awa-odoo commented on addons/website_sale_slides/static/src/xml/website_sale_slides_quiz.xml:
```diff
@@ -2,29 +2,34 @@
                     <span class="my-0 h4" style="line-height: 1">
                         <span title="Succeed and gain karma" aria-label="Succeed and gain karma" class="badge badge-pill badge-warning text-white font-weight-bold ml-3 px-2 py-1">
-                            + <t t-esc="widget.quiz.quizKarmaGain"/> XP
+                    + <t t-esc="widget.quiz.quizKarmaGain"/> XP
```
Why did you change the indentation here?

fleodoo replied:
no reason, removed it

awa-odoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -146,6 +146,16 @@ def _get_new_slide_category_values(self, channel, name):
+        for slide_id in slide_ids.ids:
+             session_slide_answer_quiz.pop(str(slide_id),None)
+        request.session['slide_answer_quiz'] = json.dumps(session_slide_answer_quiz)
+        return True
```
Don't think you ever use that return value so it seems unnecessary.

awa-odoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -754,6 +770,18 @@ def slide_quiz_submit(self, slide_id, answer_ids):
         }
+    
+    @http.route(['/slides/slide/quiz/save_slide_answsers'], type='json', auth='public', website=True)
+    def slide_save_answsers(self, quizAnswers):
```
We use only snake case for python: `quiz_answers`

awa-odoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -754,6 +770,18 @@ def slide_quiz_submit(self, slide_id, answer_ids):
+        if 'slide_answer_quiz' not in request.session:
+            request.session['slide_answer_quiz'] = json.dumps({})
+        session_slide_answer_quiz = json.loads(request.session['slide_answer_quiz'])
+        values={}
```
You need to have spaces between an operator
So `values = {}`

Your flake8 linter should warn you if correctly configured (https://github.com/odoo/enterprise/wiki/Welcome)

Please check the whole commit for those.

awa-odoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -754,6 +770,18 @@ def slide_quiz_submit(self, slide_id, answer_ids):
+        session_slide_answer_quiz = json.loads(request.session['slide_answer_quiz'])
+        values={}
+        values[str(quizAnswers['slide_id'])]= quizAnswers['slide_answers']
+        session_slide_answer_quiz.update(values)
```
Should be equivalent to 
```
session_slide_answer_quiz.update({
    str(quiz_answers['slide_id']): quiz_answers['slide_answers']
})
```

awa-odoo commented on addons/website_slides/data/slide_slide_demo.xml:
```diff
@@ -103,6 +103,98 @@
+            <field name="question_id" ref="slide_slide_demo_0_5_question_1"/>
+        </record>
+
+    <record id="slide_slide_demo_0_5" model="slide.slide">
```
Looks like you define all demo data twice.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -14,8 +14,9 @@ var CourseJoinWidget = publicWidget.Widget.extend({
     },
 
-    init: function (parent, channelId){
+    init: function (parent, channelId, quiz = false) {
```
The preferred way of handling parameters in the widgets `init` method is passing a "options" parameter.
So the signature is always: `init: function (parent, options)`

Where options is a json object of various optional parameters.

Then you do:
```
this.channelId = options.channelId;
this.quiz = options.quiz || false;
```

Can you modify the signature and adapt the calls?
(I know it was already like this before your changes, but since you're altering it and the PR targets master, we might as well fix it).

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -49,6 +50,13 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+    },
+
+    /**
+     * @public
```
Missing `@param` for channelId.

This stuff is important because the JS doc is run through a program that builds a HTML file with it.
If there are errors, the program will fail (and the JS team will be very very unhappy).

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -67,39 +72,24 @@ odoo.define('website_slides.quiz', function (require) {
+                if (self.isMember && !self.publicUser && self.slide.answerIds) {
+                    self._onSubmitQuiz();
+                }
+                self.slide.answerIds = false;
```
In which cases do you need to set this to false manually?

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -190,6 +188,23 @@ odoo.define('website_slides.quiz', function (require) {
+        _renderJoinWidget: function () {
+            var courseJoinWidget = new CourseJoinWidget(this, this.channel.channelId, true);
+            if (!this.isMember && !this.publicUser) {
+                if (this.$(".o_wslides_course_join_widget")[0]) {
```
It will work the same way but we prefer `this.$(".o_wslides_course_join_widget").length !== 0`

It's more explicit.

fleodoo commented:
@awa-odoo except for your question, I corrected everything :)

awa-odoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -754,6 +769,17 @@ def slide_quiz_submit(self, slide_id, answer_ids):
             'rankProgress': rank_progress,
         }
+    
+    @http.route(['/slides/slide/quiz/save_slide_answsers'], type='json', auth='public', website=True)
```
`answsers` -> `answers`

Please grep it to change everywhere.

I noticed you have a few spelling mistakes in your code and in your commit message.
I personally use a spell checker in my VS Code extension ("Code Spell Checker") and it helps for that kind of things.
It shows up like that for e.g:
![Selection_309](https://user-images.githubusercontent.com/41282536/67264627-f2b45400-f4ab-11e9-8b13-623e377a7fb5.png)

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -190,6 +188,23 @@ odoo.define('website_slides.quiz', function (require) {
+
+        /*
+         * @private
+         * render the button to join a cours
```
`cours` -> `course`

dbeguin commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -40,26 +101,55 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+            this.$('a.o_wslides_js_course_join_link').addClass('d-none');
+            this.$('b.o_wslides_js_course_public').addClass('d-none');
+        }
+    },
```
This could be done in template directly
So not sure the start overrides that calls this method are necessary.

dbeguin commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -17,23 +20,46 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+     * @override
+     */
+    init: function () {
+        var def = this._super.apply(this, arguments);
```
init is not a deferred
and does not return anything.
so you could simply do : 
```
init: function () {
    this._super.apply(this, arguments);
    doYourStuff;
}
```
Anyway, DOM modif can be done in template directly

dbeguin commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -67,39 +68,26 @@ odoo.define('website_slides.quiz', function (require) {
+                if (!self.isMember) {
+                    self._renderJoinWidget();
+                }
+                if (self.isMember && !self.publicUser && self.slide.answerIds) {
```
`else if (!self.publicUser && self.slide.answerIds) `

dbeguin commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -219,6 +239,47 @@ odoo.define('website_slides.quiz', function (require) {
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+        _alertHide: function () {
```
public or private ? :D

fleodoo commented:
@dbeguin

- I called it a 'refactoring' but it is not really a refactoring. It is all part of the IMP but after the review of awa he told me I needed to change quite a few things so I called the commit containing all those changes 'refactoring'. I squashed them together. So I hope the description of the commits are ok now. 

- I removed the demo-data I added.

- I will rename my branch put it is possible to rename a branch without creating a new PR? 

- I put everything in the templates and removed it from the JS.

- I also changed your other comments.

dbeguin commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -77,26 +147,27 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+    return {
+        courseJoinWidget: CourseJoinWidget,
+        websiteSlidesCourseJoin: publicWidget.registry.websiteSlidesCourseJoin
+    };
```
useless extra indentation

dbeguin commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -1,28 +1,65 @@
+     * @private
+     * @param {MouseEvent} event
+     */
+    _joinCourse: function () {
```
this method is useless.
Do directly `this.joinChannel(this.channelId);` inside `_onClickJoin`

dbeguin commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -39,16 +76,45 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+        var quizAnswers = this.quizWidget._getAnswers();
+        if (quizAnswers.length === self.quiz.questions.length) {
+            this.quizWidget.alertHide();
+            return this._rpc({
```
no need to return

dbeguin commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -77,23 +141,25 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+            });
+            return Promise.all(proms);
+        },
+    });
```
still useless extra indentation

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -350,21 +377,60 @@ odoo.define('website_slides.quiz', function (require) {
-                this._submitQuiz(values);
-            } else {
-                this._alertShow();
+            else if (alert_code === 'slide_quiz_done') {
```
You should indent your `else if` blocks like the following:

```
if (condition1) {
    // do stuff
} else if(condition2) {
    // do other stuff
}
```

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -392,6 +458,7 @@ odoo.define('website_slides.quiz', function (require) {
                 var channelData = self._extractChannelData(slideData);
                 slideData.quizData = {
                     questions: self._extractQuestionsAndAnswers(),
+                    answerIds: slideData.answerIds || false,
```
`|| false` again.

If your answerIds is an array, you should use `|| []` here, or just no default value if you always expect something to be passed.

awa-odoo commented on addons/website_slides/static/src/xml/channel_management.xml:
```diff
@@ -1,12 +1,55 @@
-                <span class="cta-title text_small_caps">Join Course</span>
-            </a>
+    <t t-name="slide.quiz.join">
+        <div id ="join_widget" class="alert alert-info d-flex align-items-center justify-content-between">
```
Should not use `id` if you can avoid it.
If it's absolutely necessary, please namespace it correctly.

fleodoo commented on addons/website_sale_slides/models/sale_order.py:
```diff
@@ -29,6 +29,6 @@ def _action_confirm(self):
 
         for sale_order, channels in channels_per_so.items():
-            channels._action_add_members(sale_order.partner_id)
+            channels.sudo()._action_add_members(sale_order.partner_id)
```
Currently, When the user has the right in sales to edit his own documents, or all documents, but has no rights on the website he is able to validate the payment for a course, put this will not put the user who bought the course in the
attendees of the course, as he could not modify those.
so I added the sudo to make sure that whenever a sale is confirmed, the course will be given to the payer.

awa-odoo replied:
Does this issue only occur in your branch?

If not, it should be extracted to another FIX branch based on the related stable branch (saas-12.3 or 13.0 depending on when the issue appeared first).
You will also have to create a separate task in our project, explain the use case, and make it tested separately.

It's a good thing to fix the issues you encounter, but we have to make sure they're fixed for older branches too.
So the clients that are in production on these older branches will receive the fix as well.

If that's unclear, don't hesitate to ask.

fleodoo replied:
No it's clear, i'll have a look at saas-12.3 and 13.0 :p

fleodoo replied:
Bon bein, I don't have this issue anymore neither on my branch neither on the other ones so I removed the sudo

fleodoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -547,6 +557,11 @@ def slide_view(self, slide, **kwargs):
+        if 'slide_answer_quiz' in request.session:
+            slide_answer_quiz = json.loads(request.session['slide_answer_quiz'])
+            if str(slide.id) in slide_answer_quiz:
+                values['answer_ids'] = json.dumps(slide_answer_quiz[str(slide.id)])
```
it indeed works without the json.dumps

fleodoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -1,28 +1,59 @@
+        this.channelEnroll = options.channel.channelEnroll || false;
+        this.isMember = options.isMember || false;
+        this.publicUser = options.publicUser || false;
+        this.slide = options.slide || false;
```
I did it like this because when CourseJoinWidget is used in websiteSlidesCourseJoin options only is:
`var options = {channel: {channelEnroll: data.channelEnroll, channelId: data.channelId}};`
but when  CourseJoinWidget is used in Quiz options is:

```
            var options = {
                quiz: this.quiz,
                channel: this.channel,
                isMember: this.isMember,
                publicUser: this.publicUser,
                slide: this.slide,                
            };
```
So only channelId and channelEnroll are present in both cases.

awa-odoo replied:
That seems like a lot of additional values to pass.
That widget probably doesn't need to get the "quiz", but I let you get to my other previous comments to see why.

The rest of the options do not seem related to a quiz usage as well.
So maybe you should pass all values even if "websiteSlidesCourseJoin" ?
Not sure what behavior they imply in the join Widget.

I'll check again when you're done with other comments, I'll be able to see more clearly then.

fleodoo replied:
I think it is better now :)

fleodoo commented on addons/website_slides/controllers/main.py:
```diff
@@ -146,6 +146,15 @@ def _get_new_slide_category_values(self, channel, name):
+        slide_ids = request.env['slide.slide'].search([('channel_id', '=', channel_id),('slide_type','=','quiz')])
+        session_slide_answer_quiz = json.loads(request.session['slide_answer_quiz'])
+        for slide_id in slide_ids.ids:
+            session_slide_answer_quiz.pop(str(slide_id), None)
```
If I don't set None I have an keyerror when submitting a Quiz without previously having set answers in session['slide_answer_quiz']

fleodoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -350,21 +406,30 @@ odoo.define('website_slides.quiz', function (require) {
+        //--------------------------------------------------------------------------
+        // Public
+        //--------------------------------------------------------------------------
+        alertHide: function () {
```
indeed :) Thanks

fleodoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -77,23 +141,25 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+        start: function () {
+            var self = this;
+            var proms = [this._super.apply(this, arguments)];
+            $('.o_wslides_js_course_join').each(function () {
```
I did not find how to do it with the this.$, it was already without when I started :p

fleodoo commented on addons/website_slides/static/src/xml/channel_management.xml:
```diff
@@ -1,12 +1,55 @@
+                                Join the course to take the quiz and verify your answers!
+                            </t>
+                        </b>
+                        <span class="my-0 h4 o_wslides_js_karma_free">
```
I don't see a big change when removing the line-height: 1, but putting small makes it really small. I just removed the line-height: 1.

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -19,7 +23,48 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
+    ),
+});
+
+CourseJoinWidget.include({
```
Extract the "CourseJoinWidget" part into a "slides_course_join.js" file.
So that it reflects the split in the base website_slides module.

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -19,7 +23,48 @@ sAnimations.registry.websiteSlidesQuizNoFullscreen.include({
     xmlDependencies: (Quiz.prototype.xmlDependencies || []).concat(
-        ["/website_sale_slides/static/src/xml/website_sale_slides_quiz.xml"]
-    )
+        ["/website_sale_slides/static/src/xml/channel_management.xml"]
```
You should put the Quiz related override in the "website_sale_slides_quiz.xml" instead of re-grouping them into the channel_management one.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -1,9 +1,8 @@
 'use strict';
 
-var core = require('web.core');
+var core = require('web.core');  
```
Why the extra space?

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -77,23 +121,25 @@ var CourseJoinWidget = publicWidget.Widget.extend({
-    selector: '.o_wslides_js_course_join_link',
-
+        selector: '.o_wslides_js_course_join_link',
+    
```
Why the extra space?

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -77,23 +121,25 @@ var CourseJoinWidget = publicWidget.Widget.extend({
-
-
+    
+    
```
Why the extra space?

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -67,40 +71,29 @@ odoo.define('website_slides.quiz', function (require) {
         /**
          * @override
+         * At the end we set self.slide.answers = false so that if you retry the quiz because it was wrongly answered
+         * _renderAnswersHighlighting doesn't wrongly render the corrections. If self.slide.answers still contains value,
```
doesn't wrongly ... = correctly?

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -285,20 +285,78 @@ odoo.define('website_slides.quiz', function (require) {
+            }
+            this.$('.o_wslides_js_lesson_quiz_error span:first').text(message);
+            this.$('.o_wslides_js_lesson_quiz_error').removeClass('d-none');
+        },
```
Why did you move those 2 methods?

fleodoo commented:
@awa-odoo changes done I think

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -39,25 +73,35 @@ var CourseJoinWidget = publicWidget.Widget.extend({
+     */
+    _signInAndJoinCourse: function () {
+        var url = this._createLoginRedirectUrl();
+        this.trigger_up('before_join_course_redirect', {url: url});
```
Sorry but I changed my mind about this.

I think the parent widget should send a new option to the CourseJoinWidget: `beforeJoin` that is a Promise.
And this code could do:
`this.beforeJoin.then(function() { // compute url and redirect });`

And the parent would init with:
- If they don't need to do anything before the redirect: 
`beforeJoin: Promise.resolve()`
- If they need to do an _rpc before the redirect:
```
beforeJoin: this._saveQuizAnswers.bind(this)

...

_saveQuizAnswers: function () {
    return self._rpc(...):
}
```

I think it will be much cleaner that way.

Again, sorry I changed my mind about this thing, but it's not trivial so we should experiment a bit and see what the best solution is.

awa-odoo commented on addons/website_slides/static/src/js/slides_course_join.js:
```diff
@@ -39,25 +73,35 @@ var CourseJoinWidget = publicWidget.Widget.extend({
-                location.reload();
+            if (!data.error) {
+                if (self.isQuiz) {
+                    self.trigger_up('submit_after_join');
```
Again, just like "beforeJoin", there should be an "afterJoin".
But it does not have to be a promise, just a regular function that acts as a callback.

That way you can get rid of the "isQuiz" variable and make the join widget agnostic to its parent "type".

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -365,20 +381,55 @@ odoo.define('website_slides.quiz', function (require) {
+            if ($widgetLocation.length !== 0) {
+                courseJoinWidget.appendTo(this.$(".o_wslides_join_course_widget"));
+                if (!this.publicUser) {
+                    if (courseJoinWidget.channelEnroll === 'public' && this.slide.answers) {
```
You can join the 2 if conditions to reduce indentation.

Also, your docstring is very weak.

You can explain a little more what you do here
(join the course automatically if the user is logged in and the enroll type is public etc.)

awa-odoo commented on addons/website_slides/static/src/js/slides_course_quiz.js:
```diff
@@ -481,21 +534,44 @@ odoo.define('website_slides.quiz', function (require) {
-
-            if (values.length === this.quiz.questionsCount){
-                this._submitQuiz(values);
+        _onSubmitQuiz: function () {
```
"on" suggests an event was triggered, but it's not the case anymore.

I think you could easily merge `_onSubmitQuiz` and `_submitQuiz`.

fleodoo replied:
It is still triggered by click .o_wslides_js_lesson_quiz_submit, should I try to not use _onSubmitQuiz in all other cases ?

awa-odoo replied:
Then yes other calls not triggered by the click event should go to `_submitQuiz` directly.

awa-odoo commented on addons/website_slides/static/src/xml/slide_quiz.xml:
```diff
@@ -40,27 +40,42 @@
+            <div t-if="!widget.isMember">
+                <div class="o_wslides_join_course alert alert-info d-flex align-items-center justify-content-between">
+                    <t t-if="widget.channel.channelEnroll == 'invite'">
+                        <div>
```
```
<div t-if="widget.channel.channelEnroll == 'invite'">
... 
<div t-else="">
```

To reduce template complexity.

awa-odoo commented on addons/website_sale_slides/static/src/js/slides_course_join.js:
```diff
@@ -0,0 +1,46 @@
+            if (this.publicUser) {
+                url = this._getLoginRedirectUrlSale(url);
+            } 
+            this.trigger_up('before_join_course_redirect', {url: url});
```
This should not be needed anymore, right?

fleodoo replied:
Indeed, I forgot to change website_sale_slide, sorry.
It is done now.
